### ISSUE TITLE
Create a CupertinoTabBar

### ISSUE THREAD

**Issue by eseidelGoogle** [flutter/flutter#9926]
Customer:six was asking how they would go about implementing a UINavigationController-like transition, where the AppBar and bottom navigation bar stay fixed, and the body of the scaffold transitions (and supports back gesture).  Which is different from the full-screen transitions which Navigator supports.  customer: tunes has also built something similar, but just by disabling transitions in their Routes I believe.

Maybe one way is to Mark the AppBar and BottomNavigationBar as heros?  @xster might have thoughts.

**Comment by HansMuller** [comment]
Why not just make the body of the app a PageView or an IndexedStack? The Scaffold's back button could be wired to the PageController or the stack's index.

**Comment by eseidelGoogle** [comment]
CC @FaisalAbid @harveynick

**Comment by eseidelGoogle** [comment]
@HansMuller I'm not sure either of those would play nice with the back gesture.  But presumably one could build a back-gesture aware container widget to put around the IndexedStack, etc.

**Comment by HansMuller** [comment]
Yes. Or you might find that the paging is an intuitive-enough way to navigate between "screens".

**Comment by xster** [comment]
I did a bit of reading. I may not fully understand all the nuances yet but I propose that we split this problem into 2 distinct parts based on I think the most important aspect which is is the bottom tab bar higher or lower than the scaffold and the navigator in the widget hierarchy.

## In the context of Material Design

As far as I can find from the patterns shown in https://material.io/guidelines/patterns/navigation.html and looking at various Android apps, bottom bars are children of the scaffold and the navigator.

There are never cases where tabs in the bottom tab bar (or top tab bars or drawers for that matter) that swaps between multiple parallel instances of navigator history stacks which all have its own states. 

For instance, Gmail, Photos, YouTube all have drawers/bottom tabs/top tabs, but only on the root level of the single navigator stack. Once you go down, the drawer, bottom tabs, top tabs aren't accessible anymore until you pop back to the root. Also, once you start going down, it's fine that the page routes etc lets you start fresh since the next page may or may not have another scaffold or another bottom tab bar.

So I think in the case of lib/src/material, the structure of MaterialApp>Navigator>Scaffold>BottomNavigationBar is already the correct structure conforming to Material Design. What the customer is describing 'shouldn't be possible' in the realm of Material. 

## In the context iOS

For iOS, the pattern is different. The correct pattern should rather be MaterialApp (or whatever)>BottomNavigationBar(or probably a CupertinoBottomTabBar extracted from BottomNavigationBar without the Material stuff)>Navigator>Scaffold (or something else).

Because 1, once the bottom tab bar is put in, it's never invisible again in any of its (potentially many) children stacks. In other words, each tab's 'child', if it were a scaffold again, shouldn't get to choose whether to build another bottom tab bar in the scaffold again. It's always there and should be unknown to the child scaffold. There also may not be scaffolds in some of the tabs. Like the iOS app store app where one of the tabs is a search field. 

The iOS bottom bar also owns multiple stacks navigator stacks. Each can have its own state and be deeply navigated as the user swaps between them.  

So in summary, I think the material package is right as is and we should probably make another CupertinoBottomTabBar and recommend the pattern be to let the user have another Navigator as first child of each tab.

**Comment by xster** [comment]
@cbracken, @abarth any thoughts? Perhaps extract the fixed part of material BottomNavigationBar into something reused by CupertinoBottomTabBar. 

Might be easier for the user too if CupertinoBottomTabBar just automatically came with a PageView that mapped to the buttons. Not sure if new API is needed or if it can be made unscrollable currently.

**Comment by xster** [comment]
Another nit: iOS app bar and bottom tab bar can have contents go under them with a blur mask as well

**Comment by xster** [comment]
This is done. Pushing other navigation pattern demo parts to #10304
